File: Difficulty/Medium/[152] Maximum Product Subarray/solution.cpp
/*
 * @lc app=leetcode id=152 lang=cpp
 *
 * [152] Maximum Product Subarray
 */

// @lc code=start
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.empty()) return 0;
        int r = nums[0];
        int imax = r, imin = r;
        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] < 0)
                swap(imax, imin);
            imax = max(nums[i], imax * nums[i]);
            imin = min(nums[i], imin * nums[i]);
            r = max(r, imax);
        }
        return r;
    }
};
// @lc code=end

File: Difficulty/Medium/[152] Maximum Product Subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums: return 0
        r = nums[0]
        imax, imin = r, r
        for i in range(1, len(nums)):
            if nums[i] < 0:
                imax, imin = imin, imax
            imax = max(nums[i], imax * nums[i])
            imin = min(nums[i], imin * nums[i])
            r = max(r, imax)
        return r